src/Workspaces/Core/Portable/Workspace/Solution/Project.cs

```diff
@@ -201,6 +201,11 @@ internal Project(Solution solution, ProjectState projectState)
     /// </summary>
     public IEnumerable<AnalyzerConfigDocument> AnalyzerConfigDocuments => AnalyzerConfigDocumentIds.Select(GetAnalyzerConfigDocument)!;
 
+    /// <summary>
+    /// The current version of source generator execution that we're on. <inheritdoc cref="CodeAnalysis.SourceGeneratorExecutionVersion"/>.
+    /// </summary>
+    internal SourceGeneratorExecutionVersion SourceGeneratorExecutionVersion => this.State.Attributes.SourceGeneratorExecutionVersion;
```

Comment:
Core new concept.  In 'automatic' mode this does nothing (we just always regenerate SGs on any change).  In 'balanced' mode we regenerate when this version changes.

Reply:
SGTM.  i will make it so you can rerun in either case.